concent_api/core/tests/test_core_views.py

```diff
@@ -496,6 +510,98 @@ def test_send_should_return_http_202_if_task_to_compute_deadline_is_correct(self
 
             self.assertIn(response_202.status_code, [200, 202])
 
+    @freeze_time("2017-11-17 10:00:00")
+    def test_send_task_to_compute_without_public_key_should_return_http_400(self):
+        assert StoredMessage.objects.count() == 0
+
+        compute_task_def = message.ComputeTaskDef()
+        compute_task_def['task_id']     = 'ABC00XYZ'
+        compute_task_def['subtask_id']  = 'ABC00XYZ'
+        compute_task_def['deadline']    = self.message_timestamp
+
+        task_to_compute                 = message.TaskToCompute(
+            compute_task_def     = compute_task_def,
+            provider_public_key  = PROVIDER_PUBLIC_KEY,
+            requestor_public_key = REQUESTOR_PUBLIC_KEY,
+        )
+
+        for field_name in [
```

Comment:
I would prefer to use:
```
    @freeze_time("2017-11-17 10:00:00")
    @pytest.mark.parametrize('attribute', [
        'provider_public_key',
        'requestor_public_key',
    ])
    def test_send_task_to_compute_without_public_key_should_return_http_400(self, attribute):
        assert StoredMessage.objects.count() == 0
        setattr(self.task_to_compute, attribute, None)
        ...
```
instead of `for` loops in the tests ;)

Reply:
Right, but we don't use pytest for now, if it should be added lets create separate issue for adding it and changing all tests like this at once, because there are more of them.

Reply:
Why not changing them gradually. Right now, changing these two tests is like ~0 effort. Having a separate task for changing everything at once might:
1) be time consuming
2) has low priority
3) introduce some bugs (the smaller the change, the easier it is to control ;)

Reply:
This is also answer for https://github.com/golemfactory/concent/pull/183#discussion_r175580697. I am referring to this part of code:

```
response = self.send_force_report_computed_task(self.task_to_compute)
self.assert_bad_request(response)
```

> introduce some bugs (the smaller the change, the easier it is to control ;)

Doing it partially drives to a different issue - lack of consistency. Currently we agreed on some level of granularity in tests, and we try to keep it everywhere, so it is easier for everyone involved to quickly read a single test or modify a bunch of them, because they are built from similar steps. When we want to change something like this, we should change it globally. There is still some mess because not all test files were refactored to use our custom test methods - and `test_core_views.py` is an example of that. So currently we have already two levels of abstractions in tests - lets not add third to it. Besides, our current custom test methods have the right level of flexibility, and adding/removing them is quite easy. I don't think we want a huge test functions that have a lot of parameters and does 90% of test job, because that can lead to errors too, and in quickly changing environment/requirements it is easier to add those small functions instead of modifying big ones in a way they still cover and pass for all cases.

> be time consuming
I think it takes more time if you sum time spent on modifying tests one by one from time to time instead of changing them all at once.

> has low priority
That is actually our internal case, so I would not treat this as a problem.

One more thing about `test_core_views.py` itself. I think there were first ever tests in this project, and there were two tasks for them:
- Test if api and api views wrapper is working correctly.
- Test partially request and responses from first use case (report computed task).
After some time, I see them as a good candidate to get rid of. All cases from first use case are covered by first use case tests `test_integration_report_computed_task` and `test_auth_report_computed_task`. Some others test validation, some other other different things. I found them difficult to maintain and they need some general cleaning, which might be done as a separate issue.

About pytest, feel free for adding issue to add pytest and modify existing tests with it.

Reply:
I talked about this with @dybi. For the record, here's my opinion:
- This improvement seems contentious so let's do it in a separate pull request like @rwrzesien suggests not to stall fixing of the original issue.
- This requires adding a new dependency (`pytest`) so in my opinion this does not count as a simple refactoring. But I am not against small refactoring in general as long as it is not mixed into the same commits with more relevant changes.
- I talked with @dybi and I think that his concern with putting off refactoring is valid but I also share @rwrzesien's concerns about this leading to inconsistent patchwork of various approaches in the code. We need a compromise. I think that our move towards smaller and quicker pull requests might be the solution here. Being able to quickly refactor something small in the whole module and have it merged in a reasonable time sounds like a good middle ground to me.
- @dybi seems averse to the overhead created by adding new issues but I'm with @rwrzesien on this one. Yeah, it does add overhead but this is the cost of being able to quickly drop or put off part of work if it's less important and takes much more time than expected. Not everything can be done perfectly in the time we have so sometimes some specific piece of refactoring suffers but it's fine as long as a reasonable amount of refactoring is getting done as we go along.
- We need more smaller, non-integration tests. @dybi calls them unit tests, I'd say functional tests because to me a pure unit test means something that's mocked-out to the point of being useless but anyway, we mean the same thing. Every public function should have a set of small tests that check just the code inside it and the private code it calls, in isolation from other, unrelated parts of the system. Integration tests should come after that and not necessarily deal with all the tiny corner cases, because as we all see they're pretty costly to maintain and add significant overhead. I think that the decision to focus on them was a good one given the circumstances but now we're quickly getting to the point were we add integration tests for things that could be more efficiently tested with unit/functional tests. `test_core_views.py` was meant to contain such tests for `core.views`, though, like other parts of the code, it went its own way. At this point it's probably better to throw this file away and start from scratch but we do need this type of tests in general.